### ISSUE TITLE
[RFC] DMD2 loss and standalone configuration

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#520]
## Feature request

Implement the **DMD2 loss and standalone DMD configuration** for the Qwen-Image MVP in #519. This replaces this issue's former generic-controller/immutable-plan scope, following [the review of #545](https://github.com/verl-project/verl-omni/pull/545#issuecomment-5609385564).

This is a CPU-verifiable foundation for the concrete engine/trainer and Qwen integration in #535/#543. It does not introduce a production trainer shell, a four-recipe registry, or claim runnable training by itself. Existing #545 code still follows the previous design until it is updated; prior passing tests do not certify this revision.

## Motivation

Separate **algorithm identity**, **shared mathematics**, and **execution**:

- Original DMD (`dmd`, deferred) combines distribution matching with paired teacher-trajectory regression.
- DMD2 (`dmd2`, MVP) removes that paired regression and trains the student against frozen real-score and trainable fake-score predictions with two-timescale optimization.
- Fake-score denoising MSE is not original DMD's paired generator-regression term.
- A class named `DMDLoss` represents the shared distribution-matching term, not proof that original DMD is implemented.

Use the existing loss registry and configuration conventions rather than creating parallel objective, recipe, rollout and initialization registries. Prompt-only data must not be silently accepted as original-DMD regression data, and `dmd2` must not be aliased to `dmd`.

### 1. Registered loss and pure tensor helpers

Add `DMDLoss(DiffusionLossFn)` to `verl_omni/trainer/diffusion/diffusion_algos.py`, using the existing `register_diffusion_loss`, `DiffusionLossResult`, validation and worker loss-dispatch conventions. Register the DMD2 path explicitly and update `DiffusionLossConfig` validation consistently.

Retain the pure functions actually used by Qwen DMD2 from `verl_omni/trainer/diffusion/distillation/utils.py`. Do not rename sound numerical code merely to reduce file count. Do not retain a universal distillation package as a prerequisite for calling these functions.

`DMDLoss` computes the student distribution-matching term. Reuse the existing pure fake-score MSE helper through the DMD engine's loss/result integration; neither loss function owns a model, scheduler, optimizer, data iterator or Ray lifecycle. The engine in #535 owns the student/fake computation branch.

### 2. Exact MVP mathematics

All tensors below describe the **same normalized Qwen latent space**. `x_g` retains the student graph; score-model evaluation is detached. For Qwen rectified-flow prediction:

```text
x_sigma = (1 - sigma) * stop_gradient(x_g) + sigma * epsilon
v_target = epsilon - stop_gradient(x_g)
x0 = x_sigma - sigma * v
```

Teacher and fake score see the same `x_sigma`, sigma and score noise in the student loss. Define under no-grad:

```text
normalizer = mean(abs(stop_gradient(x_g) - x0_real), non-batch dims, keepdim=True)
g = nan_to_num((x0_fake - x0_real) / clamp_min(normalizer, normalization_epsilon))
target = stop_gradient(x_g - g)
L_DM = 0.5 * mean((x_g - target) ** 2)
L_fake = mean((v_fake - v_target) ** 2)
```

The student objective for the MVP is `L_DM`, with an explicit configured weight if needed. It has no LPIPS/trajectory-regression or adversarial term. Original DMD would require a separate full objective such as `L_DM + lambda_reg * L_paired_teacher_regression`; that extension is not delivered by this issue.

Required numerical boundaries:

1. Score conversion, sigma arithmetic, normalization and loss reduction use fp32 for this Qwen MVP even when denoiser forwards use bf16/fp16. This is not a universal precision contract for future CausVid/Self-Forcing implementations.
2. The normalizer spans **all non-batch dimensions of each sample**, not a global batch norm and not only masked elements. Detach it and both score predictions from the student loss graph.
3. The fake-stage target is detached, and #535 must also detach samples before constructing fake-stage model inputs. Detaching only the target is insufficient.
4. Count non-finite score-gradient replacements before sanitizing them. `nan_to_num` is not evidence that the forward or optimizer update was numerically healthy.
5. Keep any retained mask helper's existing semantics: expand prefix/singleton masks to the prediction shape before counting active scalar elements; reject incompatible/all-masked inputs; a loss mask never narrows the DMD normalizer. The image MVP does not introduce causal/video masks.
6. Loss weights and microbatch/DP normalization are applied exactly once. Reusing `workers/utils/losses.py` must not divide again by an accumulation count after the engine has already applied the intended weighting.

**Gradient-test convention:** finite-difference the surrogate against a fixed detached target. Recomputing the stop-gradient target during numerical perturbation is not the derivative being checked. Assert the analytic gradient, including the mean-reduction denominator, and assert no teacher/fake-score gradients from the student loss.

### 3. Qwen guidance and sampling utilities

Retain standard CFG used by this MVP:

```text
v_cfg = v_negative + guidance_scale * (v_positive - v_negative)
```

Teacher guidance uses an explicit negative condition; student and fake-score predictions are conditional-only in the baseline. Support the existing tested normalization behavior where used: `none`, last-dimension `layer_norm` with its zero-norm boundary, and batch-global `scalar` scaling. Scalar CFG is not sample-separable, so batching parity tests must not equate it with per-sample normalization.

Reuse the tested flow `velocity_to_x0`, deterministic Euler step and timestep-shift utilities needed by #543. Full backward simulation, prompt encoding and distributed RNG decisions belong to the engine/adapter, not the loss module. Consistency re-noising, legacy video CFG, ODE regression and GAN helpers are not new MVP deliverables merely because they existed in a prototype.

### 4. Independent public configuration

Add `DiffusionDMDConfig` in `verl_omni/workers/config/diffusion/dmd.py`, with a sibling top-level Hydra **`dmd:`** group. Follow existing `BaseConfig`, field comments, `__post_init__` validation, `__all__`, YAML `_target_` and `omega_conf_to_dataclass` conventions. Reuse existing model, FSDP, optimizer and checkpoint configuration types rather than inventing DMD copies.

The following is a **proposed target configuration**, not a command supported by current #545/#546 heads:

```yaml
algorithm:
  trainer_type: distribution_matching
  sample_source: offline

actor_rollout_ref:
  model:
    algorithm: dmd2
  actor:
    diffusion_loss:
      loss_mode: dmd2

dmd:
  fake_update_ratio: 2
  teacher_guidance_scale: 4.0
  cfg_norm: layer_norm
  negative_prompt: " "
  normalization_epsilon: 0.000001
  export_role: student
```

`distribution_matching` selects the execution family; `dmd2` selects the implemented algorithm. The group name `dmd:` is a namespace, not an alias for original DMD. No `recipe`/`profile` combination is needed to distinguish DMD1 from DMD2 in this MVP.

Keep original OPD fields (`distillation.enabled`, `teacher_models`, `teacher_key`, resource settings and `scheduler`) unchanged. DMD2 launch instructions do not require users to configure that group. Preflight rejects OPD+DMD2 activation, including `distillation.enabled` or actor `use_distill_loss`, rather than starting teachers or mixing losses accidentally.

Configuration ownership:

| Settings | Owner |
| --- | --- |
| Model/checkpoint path, architecture, LoRA and FSDP execution | Existing model/engine configuration |
| Global training batch, training budget, save/resume and profiling | Existing `data` / `trainer` / profiler configuration |
| Student optimizer | Existing actor optimizer configuration |
| Fake-score optimizer | `dmd` field using the existing optimizer config type |
| Teacher guidance, score sampling, update ratio and EMA policy | `DiffusionDMDConfig` |
| Per-role microbatch overrides, if needed | DMD configuration translated into existing worker batching; no new batching implementation or new PPO-named DMD fields |
| Student/EMA export choice | `dmd.export_role`, default `student` |

Expose only fields exercised by Qwen. Validate positive non-boolean update counts, finite numeric settings, sigma bounds, CFG/export enums and required teacher conditioning; do not silently coerce invalid values. Unknown original-DMD regression, adversarial, causal or standalone-placement options fail closed. Complete production routing with #535/#543, not a nonfunctional trainer shell.

### 5. Scope removed from the old proposal

Do not require `DistillationPlan`, generic role/transport/export specifications, `UpdateSchedule`, recipe factories, `DistillationTrainerController`, completed-step hooks or a fake production `DistillationRayTrainer`. Move useful CPU fakes to tests and retain only helpers used by the concrete implementation.

The old rule that a numerical skip is a fatal zero-progress cycle is superseded by #535. The proposed MVP uses completed-cycle `global_step` and separate successful student/fake optimizer counters. This issue must not recreate a generic controller to implement that small loop.

## Tests and acceptance

- [ ] DMD2 loss registration and `DiffusionLossResult` integration work; `dmd` is not silently redirected to DMD2.
- [ ] DMD gradient sign, per-sample unmasked normalizer, finite-difference gradient and detached targets/scores pass CPU tests.
- [ ] Retained mask helpers preserve equivalent expanded masks and reject invalid/all-masked shapes.
- [ ] Teacher CFG normalization tests include zero conditional norm and global-scalar versus sample-wise behavior.
- [ ] Flow conversion, fixed-shift Euler utilities and precision boundaries match the chosen Qwen fixtures.
- [ ] Physical-batch versus microbatch loss/gradient tests catch unequal-last-microbatch weighting and double scaling.
- [ ] Typed/Hydra config tests accept the new `dmd:` group, preserve OPD configurations and reject incompatible algorithm fields; full runtime routing is tested in #535.
- [ ] Existing losses and OPD tests pass; config exports and generated YAMLs are updated and checked.
- [ ] Math tests require no model weights or GPU. New result tests distinguish numerical skips from malformed execution, without recreating the old controller hierarchy.

## Your contribution

Coordinate the revised #545 scope with #535/#543 so the series delivers one runnable Qwen DMD2 MVP. These are acceptance requirements for the new design, not inherited pass counts from the previous controller implementation. Implementation PRs must record actual test commands/results and satisfy repository human-review and AI-assistance disclosure requirements.

References: [parent RFC #519](https://github.com/verl-project/verl-omni/issues/519), [reviewer proposal](https://github.com/verl-project/verl-omni/pull/545#issuecomment-5609385564), [DMD](https://arxiv.org/abs/2311.18828), [DMD2](https://arxiv.org/abs/2405.14867), and [pinned LightX2V math](https://github.com/ModelTC/LightX2V/blob/2a603235e07ed281879ae90c5624abc05a14c35b/lightx2v_train/lightx2v_train/trainers/dmd/math.py). Behavioral references do not authorize copying incompatibly licensed source.

**Comment by OnePunchMonk** [comment]
Hi @NancyFyong, happy to pick the PR 1 scope of this RFC. Thank you!
